Disk docs (#956)
memgraph/docs: docs/reference-guide/storage-modes.md

sidebar_label: Storage modes
---

Memgraph supports three different storage modes:
* `IN_MEMORY_TRANSACTIONAL` - the default database mode that favors
    strongly-consistent ACID transactions using WAL files and snapshots,
    but requires more time and resources during data import and analysis.
* `IN_MEMORY_ANALYTICAL` - speeds up import and data analysis but offers no ACID
    guarantees besides manually created snapshots.

> * `ON_DISK_TRANSACTIONAL` - supports ACID properties in a same way as `IN_MEMORY_TRANSACTIONAL`
>     with the additional ability to store data on cheap media (e.g. HDD and SSD) with a lower
>     performance than other two storage modes. **Experimental**

Reviewer: ```suggestion
* `ON_DISK_TRANSACTIONAL` - supports ACID properties in the same way as `IN_MEMORY_TRANSACTIONAL`
    with the additional ability to store data on disk (HDD or SSD) thus trading performance for lower costs. **Experimental**
```

[`STORAGE_MODE` permission](/reference-guide/auth-module.md) to change the
storage mode.

> Switching from the in-memory storage mode to disk mode is allowed only when
> the database is empty and there is a single session active. The intention here
> is to allow switching to users only upon starting Memgraph, not later. Switching from
> disk mode to in-memory is forbidden since data may not fit into the memory.
> 
> To switch to disk storage mode the following query can be used:
> 
> ```cypher
> STORAGE MODE ON_DISK_TRANSACTIONAL;
> ```

Reviewer: ```suggestion
Switching from the in-memory storage mode to the on-disk storage mode is allowed when there is only one active session and the database is empty. As Memgraph Lab uses multiple sessions to run queries in parallel, it is currently impossible to switch to the on-disk storage mode within Memgraph Lab. You can change the storage mode to on-disk transactional using `mgconsole`, then connect to the instance with Memgraph Lab and query the instance as usual.

When querying the database, all the graph objects used in a transaction need to be able to fit in the available RAM, otherwise the database will raise an exception. 

To change the storage mode to `ON_DISK_TRANSACTIONAL`, use the following query:

```cypher
STORAGE MODE ON_DISK_TRANSACTIONAL;
```

Changing the storage mode from on-disk to in-memory is forbidden since data may not fit into the memory.

Reviewer: @as51340  be carfeul when merging this, i don't know why it doesn't recognize the last ``` i used to close the suggestion tag

Manual snapshots are created by running the `CREATE SNAPSHOT;` query. When the
query is run in the `IN_MEMORY_ANALYTICAL` mode, Memgraph guarantees that it
will be **the only** transaction present in the system, and all the other
transactions will wait until the snapshot is created to ensure its validity.

## On-disk transactional storage mode

In the on-disk transactional storage mode, disk is used as a physical storage
which allows you to save more data than the capacity of your RAM. This helps
keep the hardware costs to a minimum, but you should except slower performance
when compared to the in-memory transactional storage mode. Keep in mind that
while executing queries, all the graph objects used in the transactions still

> need to be able to fit in the RAM, or Memgraph will thrown an exception. 

Reviewer: ```suggestion
need to be able to fit in the RAM, or Memgraph will throw an exception. 
```

### Architecture

RocksDB is used as a background storage to serialize nodes and relationships
into a key-value format. The used architecture is also known as "larger than
memory" as it enables in-memory databases to save more data than the main memory
can hold, without the performance overhead caused by the buffer pool. 

The imported data is residing on the disk, while the main memory contains two
caches, one executing operations on main RocksDB instance and the other for
operations that require indices. The cache used here is Memgraph's custom

> `SkipList` cache which allows multithreaded read-write access pattern.

Reviewer: Which cache or both?

Author: SkipList is cache

Author: I mean any SkipList allows multithreaded read-write access pattern.

Reviewer: So both caches used in the main memory are SkipList?

Author: yes

Therefore, the in-memory storage mode uses a pessimistic approach and
immediately checks whether there is a conflict between two transactions. 

In the on-disk storage mode, the cache is used per transaction. This
significantly simplifies object management since there is no need to question
certain object's validity, but it also requires the optimistic approach for
conflict resolution between transactions. 

In the on-disk storage mode, the conflict is checked at the transaction's commit
time with the help of RocksDB's transaction support. This also implies that
deltas are cleared after each transaction, which can optimize memory usage
during execution. Deltas are still used to fully support Cypher's semantic of
the write queries. The design of the on-disk storage also simplifies the process
of garbage collection, since all the data is on disk.

### Isolation level

The on-disk storage mode supports only snapshot isolation level. Mostly because
it's the Memgraph viewpoint that snapshot isolation should be the default
isolation level for most applications relying on databases. But the snapshot
isolation level also simplifies the query's execution flow since there is no

> need to go on disk until the commit of the transaction.

Reviewer: What do you mean by - there is no need to go on disk ... what doesn't need to go on disk? just a bit more details to this thought?

Author: It means that data is not transferred to the disk until the commit (maybe even actual transfer to the disk happens a bit later but that is RocksDB internal).

nodes can be checked by looking only at this single node. On the other side,
uniqueness constraints require a different approach. For a node to be valid, the
engine needs to iterate through all other nodes under constraint and check
whether a conflict exists. To speed up this iteration process, nodes under
constraint are stored into a separate RocksDB instance to eliminate the cost of
iterating over nodes which are not under constraint.

### Data formats

Vertex format for main disk storage:
Key - `label1, label2, ... | vertex gid | commit_timestamp`
Value - `property1, property2`

Edge format for main disk storage:
Key - `from vertex gid | to vertex gid | 0 | edge type | edge gid | commit_timestamp`
Value - `property1, property2`
`0` is a placeholder for edge direction in future.

Format for label index on disk:

Key - `indexing label | vertex gid | commit_timestamp`

Value - `label1_id, label2_id, ... | property1, property2, ...`

Value does not contain `indexing label`.

Format for label-property index on disk:

Key - `indexing label | indexing property | vertex gid | commit_timestamp`

Value - `label1_id, label2_id, ... | property1, property2, ...`

> Value does not contain `indexing label`.

Reviewer: This requires some intro and explanation... what am I looking at?

Author: Format how data is serialized to the disk. Data = vertices and edges.

which store the exact versions of data at the specific moment in time.
Therefore, the in-memory storage mode uses a pessimistic approach and
immediately checks whether there is a conflict between two transactions. 

In the on-disk storage mode, the cache is used per transaction. This
significantly simplifies object management since there is no need to question
certain object's validity, but it also requires the optimistic approach for
conflict resolution between transactions. 

In the on-disk storage mode, the conflict is checked at the transaction's commit
time with the help of RocksDB's transaction support. This also implies that
deltas are cleared after each transaction, which can optimize memory usage
during execution. Deltas are still used to fully support Cypher's semantic of
the write queries. The design of the on-disk storage also simplifies the process
of garbage collection, since all the data is on disk.

### Isolation level

The on-disk storage mode supports only snapshot isolation level. Mostly because
it's the Memgraph viewpoint that snapshot isolation should be the default
isolation level for most applications relying on databases. But the snapshot

> isolation level also simplifies the query's execution flow since there is no
> need to go on disk until the commit of the transaction.

Reviewer: ```suggestion
isolation level also simplifies the query's execution flow since no data is transferred to the disk until the commit of the transaction.
```
 ok like this?

Author: yes

### Indices

The on-disk storage mode supports both label and label-property indices. They
are stored in separate RocksDB instances as key-value pairs so that the access
to the data is faster. Whenever the indexed node is accessed, it's stored into a
separate in-memory cache to maximize the reading speed.

### Constraints

The on-disk storage mode supports both existence and uniqueness constraints.
Existence constraints don't use context from the disk since the validity of
nodes can be checked by looking only at this single node. On the other side,
uniqueness constraints require a different approach. For a node to be valid, the
engine needs to iterate through all other nodes under constraint and check
whether a conflict exists. To speed up this iteration process, nodes under
constraint are stored into a separate RocksDB instance to eliminate the cost of
iterating over nodes which are not under constraint.

### Data formats

> 
> Vertex format for main disk storage:

Reviewer: ```suggestion

Below is the format in which data is serialized to the disk. 

Vertex format for main disk storage:
```

Key - `label1, label2, ... | vertex gid | commit_timestamp`
Value - `property1, property2`

> Edge format for main disk storage:

Reviewer: ```suggestion
Edge format for the main disk storage:
```

will be **the only** transaction present in the system, and all the other
transactions will wait until the snapshot is created to ensure its validity.

## On-disk transactional storage mode

In the on-disk transactional storage mode, disk is used as a physical storage
which allows you to save more data than the capacity of your RAM. This helps
keep the hardware costs to a minimum, but you should except slower performance
when compared to the in-memory transactional storage mode. Keep in mind that
while executing queries, all the graph objects used in the transactions still
need to be able to fit in the RAM, or Memgraph will throw an exception. 

### Architecture

RocksDB is used as a background storage to serialize nodes and relationships
into a key-value format. The used architecture is also known as "larger than
memory" as it enables in-memory databases to save more data than the main memory
can hold, without the performance overhead caused by the buffer pool. 

The imported data is residing on the disk, while the main memory contains two
caches, one executing operations on main RocksDB instance and the other for

> operations that require indices. The cache used here is Memgraph's custom
> `SkipList` cache which allows multithreaded read-write access pattern.

Reviewer: ```suggestion
operations that require indices. In both cases, Memgraph's custom
`SkipList` cache is used, which allows a multithreaded read-write access pattern.
```
